### ISSUE TITLE
Unable to compile with PGI C++ due to `auto` deduction failures

### ISSUE THREAD

**Issue by bek0s** [pybind/pybind11#1381]
Hello,

The latest free version of PGI compilers was released today. I tried to compile with it a simple pybind11 program but I get a bunch of errors. By "simple program" I mean just including `<pybind11/pybind11.h>`. Here is the output with all the errors: https://pastebin.com/jeCNBwWx
All the errors are cause by lines where something is cast to `auto`. After going through the files and manually typing the types, I was able to compile the simple example found at http://pybind11.readthedocs.io/en/master/basics.html#creating-bindings-for-a-simple-function

The changes I performed are the following:

File: `include/pybind11/detail/class.h`:
multiple lines: replace `auto` in `reinterpret_cast<detail::instance *>` lines with `detail::instance*`
517: `auto` -> `object`
554: `auto` -> `PyObject*`
561: `auto` -> `PyTypeObject*`

File: `include/pybind11/detail/init.h`:
27: `auto` -> `pybind11::detail::descr<1UL, pybind11::detail::value_and_holder>`

File: `include/pybind11/pybind11.h`:
741: `auto` -> `dict`

It appears that the PGI C++ compiler fails to deduct types when using the `auto` keyword in some places. To me, the pybind11 code I edited seems standard C++, but at the same time I am not a C++ expert. Can someone verify that the problem here is the PGI compiler and not some non-standard C++ used in pybind11?

Best Regards,
George

**Comment by jagerman** [comment]
Is it possible you don't have the right C++ level flags enabled?  `auto` is a pretty fundamental feature of C++11, but perhaps the compiler is operating in an earlier standard mode?

**Comment by jagerman** [comment]
A quick skim of the online documentation didn't reveal anything obvious, but try adding `-std=c++14` or `-std=c++11` into your compilation flags.

**Comment by bek0s** [comment]
@jagerman Thanks for the comment! The pybind11 code uses the `auto` keyword in many different places. The majority are recognized by the compiler. Only the lines I mention in my message are not recognized. Also, on my machine, I enabled C++11 and C++14 and I get the same error. Did you try to compile a simple pybind11 example with PGI C++ on your machine and compiled with no problems?

**Comment by jagerman** [comment]
I really hate the verbosity of writing:
```C++
detail::instance *instance = reinterpret_cast<detail::instance *>(nurse);
```

Have you submitted this upstream to PGI?  This is definitely a compiler bug.

With the 5 changes listed in the original report does the test suite (`make check`) pass?

**Comment by chuckatkins** [comment]
Okay, so I was able to track this down more.  It seems like the type deduction breaks down because the variable name is the same as the type trying to be deduced.  Essentially this:
```C++
class a { };

namespace foo {
namespace bar {
class b { };
void baz(a *arg) {
    auto b = reinterpret_cast<bar::b*>(arg);
}
}
}
```
The `auto` type deduction fails because `b` is both a vtype and a variable name.  So really I think all of this goes away if a variable name distinct from the type name gets used.  I.e. `auto instance = reinterpret_cast<detail::instance*>(...)` becomes `auto inst = reinterpret_cast<detail::instance*>(...)`.  It's an easy fix that still allows the auto deduction to work properly and I'd argue make the code easier to read.

I'll push up a pr soon with the necessary changes if @jagerman has no opposition to the approach.

**Comment by jagerman** [comment]
Yes, that sounds like a much better fix to me.

**Comment by ax3l** [comment]
Spoke to PGI engineers today and pointed them to Chuck's snippet.
The auto deduction issue is fixed in the upcoming 18.7 PGI compiler release.

**Comment by wjakob** [comment]
Too intrusive, unfortunately (especially since PGI is not an officially supported compiler). I'll close this for now.
